Add lowering for aten.searchsorted

benjaminglass1 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #137303
* __->__ #135701

Adds lowering for `aten.searchsorted`. This entails:

1. Adding support for multi-dimensional bucket tensors to `ops.bucketize`.
2. Adding support for striding to `ops.bucketize`.
3. Adding support for sorting tensors to `ops.bucketize`.
4. Adding a lowering for `aten.searchsorted.Tensor`.
5. Adding a basic decomposition for `aten.searchsorted.Scalar` that calls into the lowering for tensors.
6. Updating the meta-function for `aten.searchsorted` to properly check some of the sizing conditions.

Closes #135873

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

Differential Revision: [D63766514](https://our.internmc.facebook.com/intern/diff/D63766514)

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 commented:
Benchmark script is attached; outputs are below. Generally, there seem to be no time improvements relative to eager mode for the case without a sorter array, but cases with a sorter array get a ~50% speedup.

```
[----------------------------- 1-D presorted sequence ----------------------------]
                                                            |  Decomposed  |  Eager
32 threads: -----------------------------------------------------------------------
      buckets shape (32,), values shape (32, 32, 32)        |       37     |      8
      buckets shape (32,), values shape (32, 256, 32)       |       43     |     13
      buckets shape (32,), values shape (256, 32, 32)       |       44     |     13
      buckets shape (32,), values shape (256, 256, 32)      |       91     |     94
      buckets shape (64,), values shape (32, 32, 64)        |       45     |      8
      buckets shape (64,), values shape (32, 256, 64)       |       45     |     29
      buckets shape (64,), values shape (256, 32, 64)       |       44     |     30
      buckets shape (64,), values shape (256, 256, 64)      |      182     |    190
      buckets shape (128,), values shape (32, 32, 128)      |       44     |      8
      buckets shape (128,), values shape (32, 256, 128)     |       49     |     60
      buckets shape (128,), values shape (256, 32, 128)     |       49     |     60
      buckets shape (128,), values shape (256, 256, 128)    |      375     |    392
      buckets shape (256,), values shape (32, 32, 256)      |       44     |     20
      buckets shape (256,), values shape (32, 256, 256)     |      131     |    140
      buckets shape (256,), values shape (256, 32, 256)     |      131     |    150
      buckets shape (256,), values shape (256, 256, 256)    |     1032     |   1000
      buckets shape (512,), values shape (32, 32, 512)      |       52     |     57
      buckets shape (512,), values shape (32, 256, 512)     |      390     |    372
      buckets shape (512,), values shape (256, 32, 512)     |      398     |    378
      buckets shape (512,), values shape (256, 256, 512)    |     3206     |   3269
      buckets shape (1024,), values shape (32, 32, 1024)    |      144     |    160
      buckets shape (1024,), values shape (32, 256, 1024)   |     1128     |   1200
      buckets shape (1024,), values shape (256, 32, 1024)   |     1144     |   1200
      buckets shape (1024,), values shape (256, 256, 1024)  |     9187     |   9000
      buckets shape (2048,), values shape (32, 32, 2048)    |      375     |    372
      buckets shape (2048,), values shape (32, 256, 2048)   |     2992     |   2951
      buckets shape (2048,), values shape (256, 32, 2048)   |     2996     |   2952
      buckets shape (2048,), values shape (256, 256, 2048)  |    24100     |  23780

Times are in microseconds (us).

[----------------------------- 1-D unsorted sequence -----------------------------]
                                                            |  Decomposed  |  Eager
32 threads: -----------------------------------------------------------------------
      buckets shape (32,), values shape (32, 32, 32)        |       38     |     33
      buckets shape (32,), values shape (32, 256, 32)       |       44     |     41
      buckets shape (32,), values shape (256, 32, 32)       |       46     |     36
      buckets shape (32,), values shape (256, 256, 32)      |       92     |    133
      buckets shape (64,), values shape (32, 32, 64)        |       47     |     33
      buckets shape (64,), values shape (32, 256, 64)       |       49     |     67
      buckets shape (64,), values shape (256, 32, 64)       |       45     |     68
      buckets shape (64,), values shape (256, 256, 64)      |      281     |    290
      buckets shape (128,), values shape (32, 32, 128)      |       45     |     41
      buckets shape (128,), values shape (32, 256, 128)     |      108     |    134
      buckets shape (128,), values shape (256, 32, 128)     |      107     |    134
      buckets shape (128,), values shape (256, 256, 128)    |      847     |    790
      buckets shape (256,), values shape (32, 32, 256)      |       45     |     66
      buckets shape (256,), values shape (32, 256, 256)     |      318     |    324
      buckets shape (256,), values shape (256, 32, 256)     |      318     |    324
      buckets shape (256,), values shape (256, 256, 256)    |     1000     |   2285
      buckets shape (512,), values shape (32, 32, 512)      |      115     |    138
      buckets shape (512,), values shape (32, 256, 512)     |      887     |    841
      buckets shape (512,), values shape (256, 32, 512)     |      890     |    844
      buckets shape (512,), values shape (256, 256, 512)    |     3230     |   6400
      buckets shape (1024,), values shape (32, 32, 1024)    |      291     |    305
      buckets shape (1024,), values shape (32, 256, 1024)   |     1000     |   2139
      buckets shape (1024,), values shape (256, 32, 1024)   |     1000     |   2154
      buckets shape (1024,), values shape (256, 256, 1024)  |     9300     |  17000
      buckets shape (2048,), values shape (32, 32, 2048)    |      707     |    680
      buckets shape (2048,), values shape (32, 256, 2048)   |     3021     |   5273
      buckets shape (2048,), values shape (256, 32, 2048)   |     2956     |   5276
      buckets shape (2048,), values shape (256, 256, 2048)  |    23620     |  41800

Times are in microseconds (us).

[--------------------------------- N-D presorted sequence ---------------------------------]
                                                                     |  Decomposed  |  Eager
32 threads: --------------------------------------------------------------------------------
      buckets shape (32, 32, 32), values shape (32, 32, 32)          |       46     |      9
      buckets shape (32, 256, 32), values shape (32, 256, 32)        |       46     |     24
      buckets shape (256, 32, 32), values shape (256, 32, 32)        |       46     |     24
      buckets shape (256, 256, 32), values shape (256, 256, 32)      |      139     |    200
      buckets shape (32, 32, 64), values shape (32, 32, 64)          |       46     |      9
      buckets shape (32, 256, 64), values shape (32, 256, 64)        |       46     |     52
      buckets shape (256, 32, 64), values shape (256, 32, 64)        |       46     |     51
      buckets shape (256, 256, 64), values shape (256, 256, 64)      |      282     |    362
      buckets shape (32, 32, 128), values shape (32, 32, 128)        |       46     |     11
      buckets shape (32, 256, 128), values shape (32, 256, 128)      |       79     |    110
      buckets shape (256, 32, 128), values shape (256, 32, 128)      |       78     |    111
      buckets shape (256, 256, 128), values shape (256, 256, 128)    |      605     |    607
      buckets shape (32, 32, 256), values shape (32, 32, 256)        |       46     |     35
      buckets shape (32, 256, 256), values shape (32, 256, 256)      |      171     |    229
      buckets shape (256, 32, 256), values shape (256, 32, 256)      |      173     |    230
      buckets shape (256, 256, 256), values shape (256, 256, 256)    |     1343     |   1340
      buckets shape (32, 32, 512), values shape (32, 32, 512)        |       60     |     83
      buckets shape (32, 256, 512), values shape (32, 256, 512)      |      445     |    573
      buckets shape (256, 32, 512), values shape (256, 32, 512)      |      445     |    574
      buckets shape (256, 256, 512), values shape (256, 256, 512)    |     3519     |   3533
      buckets shape (32, 32, 1024), values shape (32, 32, 1024)      |      157     |    208
      buckets shape (32, 256, 1024), values shape (32, 256, 1024)    |     1212     |   1200
      buckets shape (256, 32, 1024), values shape (256, 32, 1024)    |     1212     |   1200
      buckets shape (256, 256, 1024), values shape (256, 256, 1024)  |     9751     |   9700
      buckets shape (32, 32, 2048), values shape (32, 32, 2048)      |      400     |    477
      buckets shape (32, 256, 2048), values shape (32, 256, 2048)    |     3168     |   3176
      buckets shape (256, 32, 2048), values shape (256, 32, 2048)    |     3172     |   3183
      buckets shape (256, 256, 2048), values shape (256, 256, 2048)  |    25480     |  25440

Times are in microseconds (us).

[--------------------------------- N-D unsorted sequence ----------------------------------]
                                                                     |  Decomposed  |  Eager
32 threads: --------------------------------------------------------------------------------
      buckets shape (32, 32, 32), values shape (32, 32, 32)          |       48     |     42
      buckets shape (32, 256, 32), values shape (32, 256, 32)        |       49     |     68
      buckets shape (256, 32, 32), values shape (256, 32, 32)        |       48     |     71
      buckets shape (256, 256, 32), values shape (256, 256, 32)      |      200     |    353
      buckets shape (32, 32, 64), values shape (32, 32, 64)          |       49     |     43
      buckets shape (32, 256, 64), values shape (32, 256, 64)        |       60     |    125
      buckets shape (256, 32, 64), values shape (256, 32, 64)        |       61     |    125
      buckets shape (256, 256, 64), values shape (256, 256, 64)      |      428     |    740
      buckets shape (32, 32, 128), values shape (32, 32, 128)        |       49     |     46
      buckets shape (32, 256, 128), values shape (32, 256, 128)      |      140     |    243
      buckets shape (256, 32, 128), values shape (256, 32, 128)      |      140     |    244
      buckets shape (256, 256, 128), values shape (256, 256, 128)    |      610     |   1600
      buckets shape (32, 32, 256), values shape (32, 32, 256)        |       51     |     97
      buckets shape (32, 256, 256), values shape (32, 256, 256)      |      368     |    546
      buckets shape (256, 32, 256), values shape (256, 32, 256)      |      369     |    547
      buckets shape (256, 256, 256), values shape (256, 256, 256)    |     1300     |   3922
      buckets shape (32, 32, 512), values shape (32, 32, 512)        |      130     |    200
      buckets shape (32, 256, 512), values shape (32, 256, 512)      |      573     |   1300
      buckets shape (256, 32, 512), values shape (256, 32, 512)      |      572     |   1300
      buckets shape (256, 256, 512), values shape (256, 256, 512)    |     3528     |   9968
      buckets shape (32, 32, 1024), values shape (32, 32, 1024)      |      337     |    434
      buckets shape (32, 256, 1024), values shape (32, 256, 1024)    |     1000     |   3129
      buckets shape (256, 32, 1024), values shape (256, 32, 1024)    |     1000     |   3133
      buckets shape (256, 256, 1024), values shape (256, 256, 1024)  |     9800     |  24590
      buckets shape (32, 32, 2048), values shape (32, 32, 2048)      |      859     |    960
      buckets shape (32, 256, 2048), values shape (32, 256, 2048)    |     3186     |   7390
      buckets shape (256, 32, 2048), values shape (256, 32, 2048)    |     3163     |   7400
      buckets shape (256, 256, 2048), values shape (256, 256, 2048)  |    25270     |  58800

Times are in microseconds (us).
```

Benchmark script:

```python
import torch
from torch.testing import make_tensor
from torch.utils.benchmark import Timer, Compare
import torch._inductor.config
from itertools import product
from functools import partial
import torch._dynamo.config

torch._dynamo.config.cache_size_limit = 1000
torch._inductor.config.force_disable_caches = True
torch._inductor.config.triton.cudagraphs = False

benchmark_name = "searchsorted"
XY_sizes = [32, 256]
Z_sizes = [32, 64, 128, 256, 512, 1024, 2048]


def gen_inputs():
    make_arg = partial(make_tensor, dtype=torch.float32, device="cuda:1")

    for Z, X, Y in product(Z_sizes, XY_sizes, XY_sizes):
        unsorted_seq = make_arg(Z)
        sorted_seq, sorting_indices = torch.sort(unsorted_seq)
        values = make_arg((X, Y, Z))

        # 1-D sorted case
        yield sorted_seq, values, None
        # 1-D unsorted case
        yield unsorted_seq, values, sorting_indices

    for Z, X, Y in product(Z_sizes, XY_sizes, XY_sizes):
        unsorted_seq = make_arg((X, Y, Z))
        sorted_seq, sorting_indices = torch.sort(unsorted_seq)
        values = make_arg((X, Y, Z))

        # N-D sorted case
        yield sorted_seq, values, None
        # N-D unsorted case
        yield unsorted_seq, values, sorting_indices


def benchmark(label, f, args):
    dim_label = "1-D" if len(args[0].shape) == 1 else "N-D"
    sorting_label = "presorted" if args[2] is None else "unsorted"

    # Warm up the timed function kernel.
    f(*args)
    f(*args)

    return Timer(
        "f(*args)",
        globals={"f": f, "args": args},
        label=f"{dim_label} {sorting_label} sequence",
        description=label,
        sub_label=f"buckets shape {tuple(args[0].shape)}, values shape {tuple(args[1].shape)}",
        num_threads=torch.get_num_threads(),
    ).blocked_autorange()


def compare(sorted_sequence, values, sorter=None):
    def f(s, v, sort):
        return torch.ops.aten.searchsorted(s, v, sorter=sort)

    f_compile = torch.compile(f)

    args = (sorted_sequence, values, sorter)
    yield benchmark("Decomposed", f_compile, args)
    yield benchmark("Eager", f, args)


c = Compare([r for args in gen_inputs() for r in compare(*args)])
c.trim_significant_figures()
c.print()
```

benjaminglass1 committed:
Update

[ghstack-poisoned]

Skylion007 commented on torch/_inductor/dependencies.py:
```diff
@@ -591,9 +596,25 @@ def extract_read_writes(
             inner.index_expr(name_to_index[entry.index_name], None)
-        for entry in fn.memory_usage[MemoryUsageType.BUCKETIZE]:
+
+        # At Python 3.12, this can be replaced by itertools.batched.  We need this
```
Actually this should be replace by itertools pairwise in Python 3.10, no?

benjaminglass1 replied:
@Skylion007 `pairwise` is not actually what we want here. For an iterable `A, B, C, D`, I want `(A, B), (C, D)`. `pairwise` returns `(A, B), (B, C), (C, D)`, while `batched` can be configured to return `(A, B), (C, D)`.

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 commented on test/inductor/test_torchinductor.py:
```diff
@@ -10315,6 +10315,115 @@ def fn(x):
+                sorter=sorter,
+            )
+
+        unsorted_sequence = torch.rand(3, 5)
+        sorted_sequence, sorting_indices = torch.sort(unsorted_sequence)
+        values = torch.rand(3, 7)
```
Good point, will do!

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

amjames commented on torch/_inductor/ops_handler.py:
```diff
@@ -1016,18 +1019,30 @@ def load_seed(self, name: str, offset: T):
         offsets_size: sympy.Expr,
         indexing_dtype: torch.dtype,
         right: bool,
+        offsets_indices,
```
Should probably add the correct type annotation here as the rest of the signature is annotated.

benjaminglass1 replied:
Done.

amjames reviewed (approved):
Some minor things that I caught, otherwise looks good!

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 commented on torch/_inductor/decomposition.py:
```diff
@@ -978,3 +979,68 @@ def max_pool2d_with_indices(
+    side: Optional[str] = None,
+    sorter: Optional[torch.Tensor] = None,
+) -> torch.Tensor:
+    # If the sorted_sequence is not one-dimensional, its shape must match that of vaflues
```
Done!

eellison commented on torch/_inductor/lowering.py:
```diff
@@ -2039,6 +2039,108 @@ def inner_fn(index):
+            )
+
+    else:
+        strides = sorted_sequence.get_stride()
```
We should either freeze `sorted_sequence`'s layout, or call `sorted_sequence.get_stride()` inside inner_fn

eellison reviewed (approved):
Looks good - also cc @davidberard98 to take a look who added the original bucketize impl

benjaminglass1 commented on torch/_inductor/decomposition.py:
```diff
@@ -978,3 +979,68 @@ def max_pool2d_with_indices(
+        right=right,
+        side=side,
+        sorter=sorter,
+    )[0]
```
@eellison It does exist, both as one of the cases in the `searchsorted` test I added and implicitly in the comprehensive `searchsorted` OpInfo testing.

benjaminglass1 commented on torch/_inductor/inductor_prims.py:
```diff
@@ -177,3 +177,19 @@ def _low_memory_max_pool2d_offsets_to_indices_aten(
 )
+
+
+# Custom prim used to transform searchsorted into a lowering that doesn't have keyword
+# argument tensors (disallowed by the current inductor lowering schema).
+_searchsorted_with_positional_sorter = make_prim(
```
`lowering.py:346-354`:

```python
        # kwargs tensors not supported yet unless it's a fallback op
        if not all(
            (fn in fallbacks or in_namespace(fn, "_c10d_functional")) for fn in aten_fn
        ):
            assert not any(isinstance(x, TensorBox) for x in kwargs.values())
            # explicitly assert for "out=" ops for better error messages
            assert not any(
                x == "out" for x in kwargs.keys()
            ), "out= ops aren't yet supported"
```

This is inside `register_lowering`.

eellison replied:
I think we could probably get away with removing that assertion. can do it in a follow up. there might be a couple small things but should be pretty easy to enable.

benjaminglass1 replied:
@eellison Do you expect any small things would show up in tests? Or are there any specific things I should look for when removing this?

eellison replied:
Yes would expect them to show up in tests.

benjaminglass1 replied:
@eellison I've opened a follow-on PR to see if it works or not.

benjaminglass1 commented on torch/_inductor/lowering.py:
```diff
@@ -2039,6 +2039,108 @@ def inner_fn(index):
+        tb, BackendFeature.BUCKETIZE
+    )
+    validate_strides = lambda tb: tb.get_stride()[-1] == 1  # noqa: E731
+    # sorted_sequence and sorter must both be contiguous in the last dimension for our
+    # assumptions in ops.bucketize to work.
```
Correct. I'll remove this once I get the bug fixed.

benjaminglass1 commented on torch/_inductor/decomposition.py:
```diff
@@ -978,3 +979,68 @@ def max_pool2d_with_indices(
+@register_decomposition(aten.searchsorted.Scalar)
+def searchsorted_scalar(
+    sorted_sequence: torch.Tensor,
+    self: Any,
```
Ahhh thank you I really wanted to make that more specific but couldn't find a scalar type to use.

davidberard98 commented on torch/_inductor/runtime/triton_helpers.py:
```diff
@@ -198,25 +198,38 @@ def any(a, dim):
-    indexing_dtype,
+    values: tl.tensor,
+    offsets_ptr: tl.pointer_type,
+    sorter_ptr: tl.pointer_type,  # optional pointer of the same shape as offsets_ptr, None if not present
```
tbh never seen tl.pointer_type - but should this be Optional[tl.pointer_type] ?

benjaminglass1 replied:
@davidberard98 I tried annotating it with that, but Triton apparently parses the type annotations and barfed on `Optional`, so I figured commenting about it was the next best thing.

benjaminglass1 committed:
Update

[ghstack-poisoned]

davidberard98 commented on torch/_inductor/codegen/common.py:
```diff
@@ -1770,9 +1770,12 @@ def bucketize(
         self,
         values: CSEVariable,
         offsets_name: str,
+        num_bucket_boundaries: sympy.Expr,
         offsets_size: sympy.Expr,
         indexing_dtype: torch.dtype,
         right: bool,
+        offsets_indices: CSEVariable,
+        sorter_name: Optional[str] = None,
```
is it possible to make these changes in a backward-compatible way?

problem 1: I think `torch/_inductor/jagged_lowerings.py` needs to be updated
problem 2: there are some internal lowerings that use `ops.bucketize` that would need to be updated simultaneously.

Note: if there's a reason why we can't make this backward-compatible, it's not a deal breaker (for (1) you'll just need to update the lowerings, unless I'm missing something; and for (2) I can help make the internal changes simultaneously). But if possible, can we just put `num_bucket_boundaries` at the bottom with the other new paramett and give it a default value?

cc @jbschlosser do you know if there are tests that should have caught this or did target determinator just skip them?

benjaminglass1 replied:
@davidberard98 I can look into that; I was operating under the assumption that I could see any lowerings that used this, since it was an internal API (my bad!). The main difficulty that I think we'll have with backwards compatibility is the assumption of a single dimension; having a sorter array present is trivial to do backwards compatibly.

To fix the striding issue, I was thinking about tinkering with this again, so let me give it more consideration. Maybe I can come up with a set of changes that'll be minimally disruptive, even if not fully BC.

benjaminglass1 replied:
Thanks for the pointer to `jagged_lowerings`, btw; I had indeed missed that one.

davidberard98 replied:
if you just put the new parameters at the end of the function signature and give them default arguments that match the previous behavior, would that work?

benjaminglass1 replied:
For this MR, sure. Would you be open to changing the function signature in the near future, though? There'll be at least two more parameters added for the striding, and this signature is already pretty complicated for my tastes.

benjaminglass1 replied:
After some offline discussion with @davidberard98, we've agreed that due to the striding bug, the API for `ops.bucketize` will have to change in non-backwards-compatible ways in order to force all consumers of the API to handle striding. To minimize the impact on Meta's internal work, I'll make that fix in this PR, such that only one set of changes will need to be done.

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 commented:
Updated the benchmarking script to test noncontiguous/strided tensor inputs:

```
[----------------------- 1-D presorted, contiguous sequence ----------------------]
                                                            |  Decomposed  |  Eager
32 threads: -----------------------------------------------------------------------
      buckets shape (32,), values shape (32, 32, 32)        |       37     |      9
      buckets shape (32,), values shape (32, 256, 32)       |       42     |     10
      buckets shape (32,), values shape (256, 32, 32)       |       45     |     13
      buckets shape (32,), values shape (256, 256, 32)      |       91     |     94
      buckets shape (64,), values shape (32, 32, 64)        |       45     |      9
      buckets shape (64,), values shape (32, 256, 64)       |       45     |     30
      buckets shape (64,), values shape (256, 32, 64)       |       46     |     28
      buckets shape (64,), values shape (256, 256, 64)      |      190     |    188
      buckets shape (128,), values shape (32, 32, 128)      |       47     |      9
      buckets shape (128,), values shape (32, 256, 128)     |       54     |     60
      buckets shape (128,), values shape (256, 32, 128)     |       55     |     60
      buckets shape (128,), values shape (256, 256, 128)    |      421     |    392
      buckets shape (256,), values shape (32, 32, 256)      |       45     |     21
      buckets shape (256,), values shape (32, 256, 256)     |      128     |    150
      buckets shape (256,), values shape (256, 32, 256)     |      130     |    150
      buckets shape (256,), values shape (256, 256, 256)    |     1024     |   1024
      buckets shape (512,), values shape (32, 32, 512)      |       52     |     59
      buckets shape (512,), values shape (32, 256, 512)     |      397     |    382
      buckets shape (512,), values shape (256, 32, 512)     |      403     |    384
      buckets shape (512,), values shape (256, 256, 512)    |     3198     |   3196
      buckets shape (1024,), values shape (32, 32, 1024)    |      144     |    200
      buckets shape (1024,), values shape (32, 256, 1024)   |     1123     |   1000
      buckets shape (1024,), values shape (256, 32, 1024)   |     1122     |   1000
      buckets shape (1024,), values shape (256, 256, 1024)  |     9038     |   9220

Times are in microseconds (us).

[----------------------- 1-D unsorted, contiguous sequence -----------------------]
                                                            |  Decomposed  |  Eager
32 threads: -----------------------------------------------------------------------
      buckets shape (32,), values shape (32, 32, 32)        |       38     |     32
      buckets shape (32,), values shape (32, 256, 32)       |       44     |     37
      buckets shape (32,), values shape (256, 32, 32)       |       44     |     36
      buckets shape (32,), values shape (256, 256, 32)      |       92     |    132
      buckets shape (64,), values shape (32, 32, 64)        |       47     |     34
      buckets shape (64,), values shape (32, 256, 64)       |       48     |     65
      buckets shape (64,), values shape (256, 32, 64)       |       47     |     64
      buckets shape (64,), values shape (256, 256, 64)      |      280     |    280
      buckets shape (128,), values shape (32, 32, 128)      |       46     |     40
      buckets shape (128,), values shape (32, 256, 128)     |      109     |    134
      buckets shape (128,), values shape (256, 32, 128)     |      108     |    134
      buckets shape (128,), values shape (256, 256, 128)    |      843     |    773
      buckets shape (256,), values shape (32, 32, 256)      |       48     |     64
      buckets shape (256,), values shape (32, 256, 256)     |      318     |    327
      buckets shape (256,), values shape (256, 32, 256)     |      321     |    323
      buckets shape (256,), values shape (256, 256, 256)    |     1024     |   2332
      buckets shape (512,), values shape (32, 32, 512)      |      110     |    137
      buckets shape (512,), values shape (32, 256, 512)     |      899     |    850
      buckets shape (512,), values shape (256, 32, 512)     |      904     |    860
      buckets shape (512,), values shape (256, 256, 512)    |     3196     |   6550
      buckets shape (1024,), values shape (32, 32, 1024)    |      298     |    306
      buckets shape (1024,), values shape (32, 256, 1024)   |     1100     |   2175
      buckets shape (1024,), values shape (256, 32, 1024)   |     1100     |   2130
      buckets shape (1024,), values shape (256, 256, 1024)  |     9219     |  16800

Times are in microseconds (us).

[--------------------- 1-D presorted, noncontiguous sequence ---------------------]
                                                            |  Decomposed  |  Eager
32 threads: -----------------------------------------------------------------------
      buckets shape (32,), values shape (32, 32, 32)        |       40     |     21
      buckets shape (32,), values shape (32, 256, 32)       |       46     |     26
      buckets shape (32,), values shape (256, 32, 32)       |       47     |     27
      buckets shape (32,), values shape (256, 256, 32)      |      117     |    182
      buckets shape (64,), values shape (32, 32, 64)        |       45     |     21
      buckets shape (64,), values shape (32, 256, 64)       |       48     |     54
      buckets shape (64,), values shape (256, 32, 64)       |       47     |     54
      buckets shape (64,), values shape (256, 256, 64)      |      236     |    600
      buckets shape (128,), values shape (32, 32, 128)      |       46     |     23
      buckets shape (128,), values shape (32, 256, 128)     |       69     |    110
      buckets shape (128,), values shape (256, 32, 128)     |       69     |    110
      buckets shape (128,), values shape (256, 256, 128)    |      532     |   1600
      buckets shape (256,), values shape (32, 32, 256)      |       46     |     35
      buckets shape (256,), values shape (32, 256, 256)     |      194     |    300
      buckets shape (256,), values shape (256, 32, 256)     |      197     |    300
      buckets shape (256,), values shape (256, 256, 256)    |     1565     |   4693
      buckets shape (512,), values shape (32, 32, 512)      |       71     |     84
      buckets shape (512,), values shape (32, 256, 512)     |      560     |   1678
      buckets shape (512,), values shape (256, 32, 512)     |      559     |   1677
      buckets shape (512,), values shape (256, 256, 512)    |     4455     |  13360
      buckets shape (1024,), values shape (32, 32, 1024)    |      181     |    203
      buckets shape (1024,), values shape (32, 256, 1024)   |     1423     |   4245
      buckets shape (1024,), values shape (256, 32, 1024)   |     1421     |   4214
      buckets shape (1024,), values shape (256, 256, 1024)  |    11530     |  35020

Times are in microseconds (us).

[---------------------- 1-D unsorted, noncontiguous sequence ---------------------]
                                                            |  Decomposed  |  Eager
32 threads: -----------------------------------------------------------------------
      buckets shape (32,), values shape (32, 32, 32)        |       43     |     61
      buckets shape (32,), values shape (32, 256, 32)       |       46     |     59
      buckets shape (32,), values shape (256, 32, 32)       |       49     |     61
      buckets shape (32,), values shape (256, 256, 32)      |      129     |    226
      buckets shape (64,), values shape (32, 32, 64)        |       47     |     59
      buckets shape (64,), values shape (32, 256, 64)       |       51     |     96
      buckets shape (64,), values shape (256, 32, 64)       |       50     |     96
      buckets shape (64,), values shape (256, 256, 64)      |      394     |    461
      buckets shape (128,), values shape (32, 32, 128)      |       49     |     61
      buckets shape (128,), values shape (32, 256, 128)     |      148     |    192
      buckets shape (128,), values shape (256, 32, 128)     |      149     |    189
      buckets shape (128,), values shape (256, 256, 128)    |     1158     |   1100
      buckets shape (256,), values shape (32, 32, 256)      |       55     |     85
      buckets shape (256,), values shape (32, 256, 256)     |      413     |    419
      buckets shape (256,), values shape (256, 32, 256)     |      420     |    423
      buckets shape (256,), values shape (256, 256, 256)    |     1564     |   3013
      buckets shape (512,), values shape (32, 32, 512)      |      139     |    168
      buckets shape (512,), values shape (32, 256, 512)     |     1087     |   1000
      buckets shape (512,), values shape (256, 32, 512)     |     1102     |   1000
      buckets shape (512,), values shape (256, 256, 512)    |     4453     |   7900
      buckets shape (1024,), values shape (32, 32, 1024)    |      334     |    357
      buckets shape (1024,), values shape (32, 256, 1024)   |     1500     |   2494
      buckets shape (1024,), values shape (256, 32, 1024)   |     1000     |   2539
      buckets shape (1024,), values shape (256, 256, 1024)  |    11670     |  20000

Times are in microseconds (us).

[--------------------------- N-D presorted, contiguous sequence ---------------------------]
                                                                     |  Decomposed  |  Eager
32 threads: --------------------------------------------------------------------------------
      buckets shape (32, 32, 32), values shape (32, 32, 32)          |       45     |      9
      buckets shape (32, 256, 32), values shape (32, 256, 32)        |       46     |     24
      buckets shape (256, 32, 32), values shape (256, 32, 32)        |       46     |     24
      buckets shape (256, 256, 32), values shape (256, 256, 32)      |      138     |    164
      buckets shape (32, 32, 64), values shape (32, 32, 64)          |       46     |      9
      buckets shape (32, 256, 64), values shape (32, 256, 64)        |       46     |     51
      buckets shape (256, 32, 64), values shape (256, 32, 64)        |       46     |     51
      buckets shape (256, 256, 64), values shape (256, 256, 64)      |      280     |    359
      buckets shape (32, 32, 128), values shape (32, 32, 128)        |       47     |     11
      buckets shape (32, 256, 128), values shape (32, 256, 128)      |       78     |    110
      buckets shape (256, 32, 128), values shape (256, 32, 128)      |       78     |    110
      buckets shape (256, 256, 128), values shape (256, 256, 128)    |      601     |    601
      buckets shape (32, 32, 256), values shape (32, 32, 256)        |       46     |     34
      buckets shape (32, 256, 256), values shape (32, 256, 256)      |      170     |    229
      buckets shape (256, 32, 256), values shape (256, 32, 256)      |      171     |    228
      buckets shape (256, 256, 256), values shape (256, 256, 256)    |     1337     |   1300
      buckets shape (32, 32, 512), values shape (32, 32, 512)        |       59     |     84
      buckets shape (32, 256, 512), values shape (32, 256, 512)      |      436     |    574
      buckets shape (256, 32, 512), values shape (256, 32, 512)      |      438     |    572
      buckets shape (256, 256, 512), values shape (256, 256, 512)    |     3467     |   3476
      buckets shape (32, 32, 1024), values shape (32, 32, 1024)      |      156     |    208
      buckets shape (32, 256, 1024), values shape (32, 256, 1024)    |     1207     |   1207
      buckets shape (256, 32, 1024), values shape (256, 32, 1024)    |     1194     |   1207
      buckets shape (256, 256, 1024), values shape (256, 256, 1024)  |     9618     |   9618

Times are in microseconds (us).

[--------------------------- N-D unsorted, contiguous sequence ----------------------------]
                                                                     |  Decomposed  |  Eager
32 threads: --------------------------------------------------------------------------------
      buckets shape (32, 32, 32), values shape (32, 32, 32)          |       46     |     41
      buckets shape (32, 256, 32), values shape (32, 256, 32)        |       48     |     72
      buckets shape (256, 32, 32), values shape (256, 32, 32)        |       47     |     71
      buckets shape (256, 256, 32), values shape (256, 256, 32)      |      200     |    351
      buckets shape (32, 32, 64), values shape (32, 32, 64)          |       48     |     42
      buckets shape (32, 256, 64), values shape (32, 256, 64)        |       60     |    124
      buckets shape (256, 32, 64), values shape (256, 32, 64)        |       60     |    124
      buckets shape (256, 256, 64), values shape (256, 256, 64)      |      429     |    730
      buckets shape (32, 32, 128), values shape (32, 32, 128)        |       48     |     46
      buckets shape (32, 256, 128), values shape (32, 256, 128)      |      130     |    243
      buckets shape (256, 32, 128), values shape (256, 32, 128)      |      140     |    242
      buckets shape (256, 256, 128), values shape (256, 256, 128)    |      601     |   1600
      buckets shape (32, 32, 256), values shape (32, 32, 256)        |       52     |     96
      buckets shape (32, 256, 256), values shape (32, 256, 256)      |      368     |    537
      buckets shape (256, 32, 256), values shape (256, 32, 256)      |      369     |    536
      buckets shape (256, 256, 256), values shape (256, 256, 256)    |     1300     |   3922
      buckets shape (32, 32, 512), values shape (32, 32, 512)        |      130     |    190
      buckets shape (32, 256, 512), values shape (32, 256, 512)      |      573     |   1300
      buckets shape (256, 32, 512), values shape (256, 32, 512)      |      572     |   1300
      buckets shape (256, 256, 512), values shape (256, 256, 512)    |     3469     |   9970
      buckets shape (32, 32, 1024), values shape (32, 32, 1024)      |      337     |    419
      buckets shape (32, 256, 1024), values shape (32, 256, 1024)    |     1207     |   3012
      buckets shape (256, 32, 1024), values shape (256, 32, 1024)    |     1207     |   3010
      buckets shape (256, 256, 1024), values shape (256, 256, 1024)  |     9618     |  24150

Times are in microseconds (us).

[------------------------- N-D presorted, noncontiguous sequence --------------------------]
                                                                     |  Decomposed  |  Eager
32 threads: --------------------------------------------------------------------------------
      buckets shape (32, 32, 32), values shape (32, 32, 32)          |       46     |     21
      buckets shape (32, 256, 32), values shape (32, 256, 32)        |       46     |     47
      buckets shape (256, 32, 32), values shape (256, 32, 32)        |       46     |     47
      buckets shape (256, 256, 32), values shape (256, 256, 32)      |      196     |    400
      buckets shape (32, 32, 64), values shape (32, 32, 64)          |       46     |     22
      buckets shape (32, 256, 64), values shape (32, 256, 64)        |       52     |     98
      buckets shape (256, 32, 64), values shape (256, 32, 64)        |       52     |     98
      buckets shape (256, 256, 64), values shape (256, 256, 64)      |      396     |   1188
      buckets shape (32, 32, 128), values shape (32, 32, 128)        |       46     |     25
      buckets shape (32, 256, 128), values shape (32, 256, 128)      |      103     |    200
      buckets shape (256, 32, 128), values shape (256, 32, 128)      |      103     |    200
      buckets shape (256, 256, 128), values shape (256, 256, 128)    |      789     |   2364
      buckets shape (32, 32, 256), values shape (32, 32, 256)        |       47     |     56
      buckets shape (32, 256, 256), values shape (32, 256, 256)      |      223     |    600
      buckets shape (256, 32, 256), values shape (256, 32, 256)      |      226     |    600
      buckets shape (256, 256, 256), values shape (256, 256, 256)    |     1748     |   5264
      buckets shape (32, 32, 512), values shape (32, 32, 512)        |       78     |    130
      buckets shape (32, 256, 512), values shape (32, 256, 512)      |      590     |   1800
      buckets shape (256, 32, 512), values shape (256, 32, 512)      |      592     |   1800
      buckets shape (256, 256, 512), values shape (256, 256, 512)    |     4665     |  14000
      buckets shape (32, 32, 1024), values shape (32, 32, 1024)      |      193     |    400
      buckets shape (32, 256, 1024), values shape (32, 256, 1024)    |     1505     |   4516
      buckets shape (256, 32, 1024), values shape (256, 32, 1024)    |     1505     |   4515
      buckets shape (256, 256, 1024), values shape (256, 256, 1024)  |    12000     |  36010

Times are in microseconds (us).

[-------------------------- N-D unsorted, noncontiguous sequence --------------------------]
                                                                     |  Decomposed  |  Eager
32 threads: --------------------------------------------------------------------------------
      buckets shape (32, 32, 32), values shape (32, 32, 32)          |       49     |     71
      buckets shape (32, 256, 32), values shape (32, 256, 32)        |       49     |    140
      buckets shape (256, 32, 32), values shape (256, 32, 32)        |       48     |    140
      buckets shape (256, 256, 32), values shape (256, 256, 32)      |      320     |    868
      buckets shape (32, 32, 64), values shape (32, 32, 64)          |       49     |     71
      buckets shape (32, 256, 64), values shape (32, 256, 64)        |       87     |    261
      buckets shape (256, 32, 64), values shape (256, 32, 64)        |       87     |    260
      buckets shape (256, 256, 64), values shape (256, 256, 64)      |      631     |   1750
      buckets shape (32, 32, 128), values shape (32, 32, 128)        |       49     |     92
      buckets shape (32, 256, 128), values shape (32, 256, 128)      |      190     |    506
      buckets shape (256, 32, 128), values shape (256, 32, 128)      |      200     |    509
      buckets shape (256, 256, 128), values shape (256, 256, 128)    |      800     |   3649
      buckets shape (32, 32, 256), values shape (32, 32, 256)        |       66     |    170
      buckets shape (32, 256, 256), values shape (32, 256, 256)      |      484     |   1100
      buckets shape (256, 32, 256), values shape (256, 32, 256)      |      485     |   1100
      buckets shape (256, 256, 256), values shape (256, 256, 256)    |     2000     |   8000
      buckets shape (32, 32, 512), values shape (32, 32, 512)        |      162     |    328
      buckets shape (32, 256, 512), values shape (32, 256, 512)      |     1253     |   2318
      buckets shape (256, 32, 512), values shape (256, 32, 512)      |     1253     |   2316
      buckets shape (256, 256, 512), values shape (256, 256, 512)    |     4672     |  18000
      buckets shape (32, 32, 1024), values shape (32, 32, 1024)      |      404     |    700
      buckets shape (32, 256, 1024), values shape (32, 256, 1024)    |     1505     |   5113
      buckets shape (256, 32, 1024), values shape (256, 32, 1024)    |     1505     |   5103
      buckets shape (256, 256, 1024), values shape (256, 256, 1024)  |    12000     |  40660
```

Script:

```python
import torch
from torch.testing import make_tensor
from torch.utils.benchmark import Timer, Compare
import torch._inductor.config
from itertools import product
from functools import partial
import torch._dynamo.config

torch._dynamo.config.cache_size_limit = 1000
torch._inductor.config.force_disable_caches = True
torch._inductor.config.triton.cudagraphs = False

benchmark_name = "searchsorted"
XY_sizes = (32, 256)
Z_sizes = (32, 64, 128, 256, 512, 1024)
booleans = (False, True)


def gen_inputs():
    make_arg = partial(make_tensor, dtype=torch.float32, device="cuda:1")

    for Z, X, Y, noncontiguous in product(Z_sizes, XY_sizes, XY_sizes, booleans):
        unsorted_seq = make_arg(Z, noncontiguous=noncontiguous)
        values = make_arg((X, Y, Z), noncontiguous=noncontiguous)

        sorted_seq, sorting_indices = torch.sort(unsorted_seq)
        if noncontiguous:
            sorted_seq = torch.repeat_interleave(sorted_seq, 2, dim=-1)[..., ::2]
            sorting_indices = torch.repeat_interleave(sorting_indices, 2, dim=-1)[
                ..., ::2
            ]

        # 1-D sorted case
        yield sorted_seq, values, None
        # 1-D unsorted case
        yield unsorted_seq, values, sorting_indices

    for Z, X, Y, noncontiguous in product(Z_sizes, XY_sizes, XY_sizes, booleans):
        unsorted_seq = make_arg((X, Y, Z), noncontiguous=noncontiguous)
        values = make_arg((X, Y, Z), noncontiguous=noncontiguous)

        sorted_seq, sorting_indices = torch.sort(unsorted_seq)
        if noncontiguous:
            sorted_seq = torch.repeat_interleave(sorted_seq, 2, dim=-1)[..., ::2]
            sorting_indices = torch.repeat_interleave(sorting_indices, 2, dim=-1)[
                ..., ::2
            ]

        # N-D sorted case
        yield sorted_seq, values, None
        # N-D unsorted case
        yield unsorted_seq, values, sorting_indices


def benchmark(label, f, args):
    dim_label = "1-D" if len(args[0].shape) == 1 else "N-D"
    sorting_label = "presorted" if args[2] is None else "unsorted"
    contiguity_label = "contiguous" if args[0].stride()[-1] == 1 else "noncontiguous"

    # Warm up the timed function kernel.
    f(*args)
    f(*args)

    return Timer(
        "f(*args)",
        globals={"f": f, "args": args},
        label=f"{dim_label} {sorting_label}, {contiguity_label} sequence",
        description=label,
        sub_label=f"buckets shape {tuple(args[0].shape)}, values shape {tuple(args[1].shape)}",
        num_threads=torch.get_num_threads(),
    ).blocked_autorange()


def compare(sorted_sequence, values, sorter=None):
    def f(s, v, sort):
        return torch.ops.aten.searchsorted(s, v, sorter=sort)

    f_compile = torch.compile(f)

    args = (sorted_sequence, values, sorter)
    yield benchmark("Decomposed", f_compile, args)
    yield benchmark("Eager", f, args)


c = Compare([r for args in gen_inputs() for r in compare(*args)])
c.trim_significant_figures()
c.print()
```

benjaminglass1 committed:
Update

[ghstack-poisoned]

amjames commented on torch/_inductor/lowering.py:
```diff
@@ -2071,6 +2071,115 @@ def inner_fn(index):
+        tb, BackendFeature.BUCKETIZE
+    )
+    # sorted_sequence and sorter must both be contiguous in the last dimension for our
+    # assumptions in ops.bucketize to work.
```
Can this note be removed now?

amjames commented on torch/_inductor/runtime/triton_helpers.py:
```diff
@@ -198,25 +198,47 @@ def any(a, dim):
+    sorter_ptr: tl.tensor,  # optional pointer of the same shape as offsets_ptr, None if not present
+    SORTER_STRIDE: tl.constexpr,
+    sorter_indices: tl.tensor,
     BLOCK_SHAPE,  # tuple/list of block shape
```
Could you add type annotations for `BLOCK_SHAPE`, and `right`?

benjaminglass1 replied:
It doesn't look like I can; `triton` barfs on parsing type annotations it doesn't understand, and I can't find a suitable version of these that it does understand.

amjames replied:
What have you tried. I think `tl.constexpr` would fit here, no?

benjaminglass1 replied:
Does `tl.constexpr` apply to something with `len() > 1`?

amjames replied:
Ah! `BLOCK_SHAPE != BLOCK_SIZE` it should work for `right` though?

benjaminglass1 replied:
Weirdly no. `triton` didn't understand the `bool` type. I was just as surprised as you. :sweat_smile:

amjames replied:
Sorry, I mean giving `right` a `tl.constexpr`. Triton should be fine with that

amjames reviewed (commented):
I can give this a deeper look in a little bit, but I think its ready for @davidberard98 to take another look.

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

amjames commented:
cc @davidberard98 (for requested changes) and @eellison to give this one another look since that has been some significant refactoring.

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

eellison reviewed (commented):
already approved by me - differing to @davidberard98 on resolving comments he requested changes for

davidberard98 commented on torch/_inductor/runtime/triton_helpers.py:
```diff
@@ -198,25 +198,71 @@ def any(a, dim):
-    BLOCK_SHAPE,  # tuple/list of block shape
+    values: tl.tensor,
+    boundaries_ptr: tl.tensor,
+    BOUNDARIES_SIZE: tl.constexpr,
+    BOUNDARIES_UNDERLYING_NUMEL: tl.constexpr,
+    BOUNDARIES_STRIDE: tl.constexpr,
```
should these be constexpr?

afaict, they are not being treated as constexpr. I did a local test, and I see dynamic shapes being passed in if the offsets size changes. And my test is not failing, which I think means that these are not actually being handled as constexpr.

do you think you could remove `constexpr` if it's not needed here? And if it _is_ needed - then can you try testing to see if things work as expected?

(same applies to the other tl.constexprs below in this function)

benjaminglass1 replied:
@davidberard98 I'll remove the constexpr, and see if I can get a more appropriate annotation.

benjaminglass1 replied:
@davidberard98 Done, and rebased on the latest `viable/strict`. Should be ready to go for you to import!

davidberard98 commented on torch/_inductor/codegen/common.py:
```diff
@@ -2025,27 +2027,81 @@ def sort(
                 values: CSEVariable,
-                offsets_name: str,
-                offsets_size: sympy.Expr,
+                boundaries: Tuple[str, sympy.Expr, sympy.Expr, sympy.Expr],
```
just curious, is there a reason to provide this as a tuple instead of 4 separate args?

benjaminglass1 replied:
@davidberard98 I agonized over this interface for a while. Passing it as 4 separate args was the simplest approach, but also made the function signature really bloated. It doesn't technically make any difference, but I often find that a shorter function signature is a surprisingly good aid to developer comprehension, by appropriately grouping things together as needed.

As another way of justifying this... `boundaries` here represents a tensor, and it makes sense to me to pass pointer/size/strides as a single, tensor-lite object.

davidberard98 replied:
Sure, makes sense

davidberard98 reviewed (commented):
sorry for the delay - left a few comments, but generally LGTM at a high level.

However, since this does change some APIs that are used internally at Meta, I think I'll need to land this manually. Once the PR is ready (any PRs that it depends on are landed, plus any other comments resolved), could you let me know and I can make sure it gets landed with all the necessary Meta-only changes?

benjaminglass1 committed:
Update on "Add lowering for aten.searchsorted"


Adds lowering for `aten.searchsorted`. This entails:

1. Adding support for multi-dimensional bucket tensors to `ops.bucketize`.
2. Adding support for striding to `ops.bucketize`.
3. Adding support for sorting tensors to `ops.bucketize`.
4. Adding a lowering for `aten.searchsorted.Tensor`.
5. Adding a basic decomposition for `aten.searchsorted.Scalar` that calls into the lowering for tensors.
6. Updating the meta-function for `aten.searchsorted` to properly check some of the sizing conditions.

Closes #135873

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

benjaminglass1 committed:
Update on "Add lowering for aten.searchsorted"


Adds lowering for `aten.searchsorted`. This entails:

1. Adding support for multi-dimensional bucket tensors to `ops.bucketize`.
2. Adding support for striding to `ops.bucketize`.
3. Adding support for sorting tensors to `ops.bucketize`.
4. Adding a lowering for `aten.searchsorted.Tensor`.
5. Adding a basic decomposition for `aten.searchsorted.Scalar` that calls into the lowering for tensors.
6. Updating the meta-function for `aten.searchsorted` to properly check some of the sizing conditions.

Closes #135873

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

benjaminglass1 committed:
Update on "Add lowering for aten.searchsorted"


Adds lowering for `aten.searchsorted`. This entails:

1. Adding support for multi-dimensional bucket tensors to `ops.bucketize`.
2. Adding support for striding to `ops.bucketize`.
3. Adding support for sorting tensors to `ops.bucketize`.
4. Adding a lowering for `aten.searchsorted.Tensor`.
5. Adding a basic decomposition for `aten.searchsorted.Scalar` that calls into the lowering for tensors.
6. Updating the meta-function for `aten.searchsorted` to properly check some of the sizing conditions.

Closes #135873

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D63766514).

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D63766514).

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D63766514).
